chore: Migrate the sync and scanAccounts functions

ksiva-ledger opened the pull request:
Let me know what you think about the API functions' organization and the types I defined. For the moment I kept everything in the same file but we can split that when we'll progress with the other functionalities.

The Stellar SDK uses TypeScript, I will spend some time after merging this branch to see if we can transform those types into Flow types.

haammar-ledger reviewed (commented):
First pass, I didn't go through all api functions yet, but I left a few suggestions.

Also you can move the whole content of the `index.js` file to a `horizon.js` file, and use the `index.js` file only to export the api. You can check how it's done for Polkadot as an example.

I tested the sync, it looks good :+1: however I'm noticing the `! sum of ops` error message, do yo have it too?
Here it is for example for the accounts on the Obelix seed:
```
$ node ./bin/index.js sync -c stellar
Stellar 1: 2.3742 XLM (14ops) (GA2S6CABCO4OBFX7PQSK6WQ3ZSX3WYF52TVBXQFLJAEPKKRFUXVFJUSD on 44'/148'/0') sep5#0  (! sum of ops 2.3744 XLM)
OPERATIONS (14)
  -2.5 XLM           OUT        d16b2fe39030b5ba06fbe08b8959a2d10f5ea864f3e1ee4c1aee78043a3a104a 2020-06-16T12:52
  -0.002 XLM         OUT        c873cfddf93c8c2c3fbf96a75a540a29bf3eccab4f05406d65184daa5885de85 2020-06-12T14:01
  -0.1 XLM           OUT        a563dc39ade732d0d27c1b9861a45f4d2357d6d58969b4f8e4d2e6185867e9e9 2020-06-12T13:35
  -0.0001 XLM        OUT        904e0728b1f108a84d889ad75f3e6fe3772a3fc58f058aebc9ea27eaa30b6911 2020-04-27T13:35
  -0.0001 XLM        OUT        d8d6b33ce82ecd7d9d9fbd640f9af437f7ab7cdc92cc2b7cb99b529991220281 2020-04-16T13:29
  +3 XLM             IN         4366cffe88da0b8ff379ab1f533ae4ef48a2cc0b8fddabef65160ba546cc1724 2020-04-10T09:02
  -1 XLM             OUT        67167188f0eee703a78b1b8592dc0ffd6916b08e924c3a9a6a448743ee1cf3c7 2020-04-09T14:53
  -0.001 XLM         OUT        8ec0b206916ddd414db7eb0d1d74ef962cc14fb86a85cee924f01935614349df 2020-04-09T08:24
  -0.0001 XLM        OUT        c8f01e95f1a4b289d701f84f7cf97c8e469279a7014f0d6df368ecd6375b3884 2020-03-13T09:50
  -0.0123 XLM        OUT        4ee461c717923f744d8aff9a9426946f3fe8941515aa5c9e863e023af298c285 2020-03-05T13:20
  -0.01 XLM          OUT        9fa1cea94b11db5fc52d4077c0c69b56de5017df372fcb3356566eb3ca2361a8 2020-03-04T17:54
  -2 XLM             OUT        351362d4509d9bc3f33b39871755185b49ad5d14e45da16234438a35377e858a 2020-03-04T08:55
  -1 XLM             OUT        93e9739d77458ab76d0f017c1ac895f78f52bbd0b08c0a67acd2cfbc339ed931 2020-02-05T13:03
  +6 XLM             IN         adb53668f9d6b97663463797ee5850bed9dd68a197f60678a0ec4c01fc9bdbb7 2020-01-17T15:28
Stellar 2: 3.1254 XLM (11ops) (GBAMU3EJX6KLW2JEIAIEAYNLPKHFPJR6OYQYX5HPYB3CVQ6QD4XUJ23J on 44'/148'/1') sep5#1  (! sum of ops 3.1255 XLM)
OPERATIONS (11)
  +0.002 XLM         IN         c873cfddf93c8c2c3fbf96a75a540a29bf3eccab4f05406d65184daa5885de85 2020-06-12T14:01
  +0.1 XLM           IN         a563dc39ade732d0d27c1b9861a45f4d2357d6d58969b4f8e4d2e6185867e9e9 2020-06-12T13:35
  +0.0001 XLM        IN         904e0728b1f108a84d889ad75f3e6fe3772a3fc58f058aebc9ea27eaa30b6911 2020-04-27T13:35
  +0.0001 XLM        IN         d8d6b33ce82ecd7d9d9fbd640f9af437f7ab7cdc92cc2b7cb99b529991220281 2020-04-16T13:29
  +1 XLM             IN         67167188f0eee703a78b1b8592dc0ffd6916b08e924c3a9a6a448743ee1cf3c7 2020-04-09T14:53
  +0.001 XLM         IN         8ec0b206916ddd414db7eb0d1d74ef962cc14fb86a85cee924f01935614349df 2020-04-09T08:24
  +0.0123 XLM        IN         4ee461c717923f744d8aff9a9426946f3fe8941515aa5c9e863e023af298c285 2020-03-05T13:20
  +0.01 XLM          IN         9fa1cea94b11db5fc52d4077c0c69b56de5017df372fcb3356566eb3ca2361a8 2020-03-04T17:54
  +0 XLM             NONE       3294d225fb976707ba1680cbbb4b68ce512d99bc4c8abc2e89268716175c03a3 2020-03-04T09:27
  +0 XLM             NONE       52a6df1e2280cbde17943e52ec0dd799a96ae609b53bf81c4734703eeb27f43a 2020-03-04T09:22
  +2 XLM             IN         351362d4509d9bc3f33b39871755185b49ad5d14e45da16234438a35377e858a 2020-03-04T08:55
Stellar 3: 0 XLM (0ops) (GDYPMQMYW2JTLPWAUAHIDY3E4VHP5SGTFC5SMA45L7ZPOTHWQ2PHEW3E on 44'/148'/2') sep5#2 
OPERATIONS (0)
```

ksiva-ledger commented on src/families/stellar/api/index.js:
```diff
@@ -0,0 +1,258 @@
+  fee_charged: string,
+  ledger_attr: number,
+  source_account_sequence: number,
+};
```
Done

ksiva-ledger commented on src/families/stellar/api/index.js:
```diff
@@ -0,0 +1,258 @@
+const LIMIT = 200;
+const FALLBACK_BASE_FEE = 100;
+
+const server = new StellarSdk.Server("https://horizon.stellar.org");
```
Done

ksiva-ledger commented:
> First pass, I didn't go through all api functions yet, but I left a few suggestions.
> 
> Also you can move the whole content of the `index.js` file to a `horizon.js` file, and use the `index.js` file only to export the api. You can check how it's done for Polkadot as an example.
> 
> I tested the sync, it looks good 👍 however I'm noticing the `! sum of ops` error message, do you have it too?
> Here it is for example for the accounts on the Obelix seed:
> (EDIT: need to scroll on the right because of dumb GitHub display...)

I moved the content of the API functions into an `horizon.js` file 👍🏽 

I also had a `! sum of ops` error which I didn't catch before. I fixed it by taking into account the fees for outgoing operations.

haammar-ledger reviewed (commented):
I tested with your changes and I don't have the error anymore on the 1st account :clap: but it is still here on the 2nd one.
I think that account has something special... I don't remember what exactly I'll try to check tomorrow.
```
$ node ./bin/index.js sync -c stellar
Stellar 1: 2.3742 XLM (14ops) (GA2S6CABCO4OBFX7PQSK6WQ3ZSX3WYF52TVBXQFLJAEPKKRFUXVFJUSD on 44'/148'/0') sep5#0 
OPERATIONS (14)
  -2.50001 XLM       OUT        d16b2fe39030b5ba06fbe08b8959a2d10f5ea864f3e1ee4c1aee78043a3a104a 2020-06-16T12:52
  -0.00201 XLM       OUT        c873cfddf93c8c2c3fbf96a75a540a29bf3eccab4f05406d65184daa5885de85 2020-06-12T14:01
  -0.10001 XLM       OUT        a563dc39ade732d0d27c1b9861a45f4d2357d6d58969b4f8e4d2e6185867e9e9 2020-06-12T13:35
  -0.00011 XLM       OUT        904e0728b1f108a84d889ad75f3e6fe3772a3fc58f058aebc9ea27eaa30b6911 2020-04-27T13:35
  -0.00011 XLM       OUT        d8d6b33ce82ecd7d9d9fbd640f9af437f7ab7cdc92cc2b7cb99b529991220281 2020-04-16T13:29
  +3 XLM             IN         4366cffe88da0b8ff379ab1f533ae4ef48a2cc0b8fddabef65160ba546cc1724 2020-04-10T09:02
  -1.00001 XLM       OUT        67167188f0eee703a78b1b8592dc0ffd6916b08e924c3a9a6a448743ee1cf3c7 2020-04-09T14:53
  -0.00101 XLM       OUT        8ec0b206916ddd414db7eb0d1d74ef962cc14fb86a85cee924f01935614349df 2020-04-09T08:24
  -0.00011 XLM       OUT        c8f01e95f1a4b289d701f84f7cf97c8e469279a7014f0d6df368ecd6375b3884 2020-03-13T09:50
  -0.01231 XLM       OUT        4ee461c717923f744d8aff9a9426946f3fe8941515aa5c9e863e023af298c285 2020-03-05T13:20
  -0.01001 XLM       OUT        9fa1cea94b11db5fc52d4077c0c69b56de5017df372fcb3356566eb3ca2361a8 2020-03-04T17:54
  -2.00001 XLM       OUT        351362d4509d9bc3f33b39871755185b49ad5d14e45da16234438a35377e858a 2020-03-04T08:55
  -1.00001 XLM       OUT        93e9739d77458ab76d0f017c1ac895f78f52bbd0b08c0a67acd2cfbc339ed931 2020-02-05T13:03
  +6 XLM             IN         adb53668f9d6b97663463797ee5850bed9dd68a197f60678a0ec4c01fc9bdbb7 2020-01-17T15:28
Stellar 2: 3.1254 XLM (11ops) (GBAMU3EJX6KLW2JEIAIEAYNLPKHFPJR6OYQYX5HPYB3CVQ6QD4XUJ23J on 44'/148'/1') sep5#1  (! sum of ops 3.1255 XLM)
OPERATIONS (11)
  +0.002 XLM         IN         c873cfddf93c8c2c3fbf96a75a540a29bf3eccab4f05406d65184daa5885de85 2020-06-12T14:01
  +0.1 XLM           IN         a563dc39ade732d0d27c1b9861a45f4d2357d6d58969b4f8e4d2e6185867e9e9 2020-06-12T13:35
  +0.0001 XLM        IN         904e0728b1f108a84d889ad75f3e6fe3772a3fc58f058aebc9ea27eaa30b6911 2020-04-27T13:35
  +0.0001 XLM        IN         d8d6b33ce82ecd7d9d9fbd640f9af437f7ab7cdc92cc2b7cb99b529991220281 2020-04-16T13:29
  +1 XLM             IN         67167188f0eee703a78b1b8592dc0ffd6916b08e924c3a9a6a448743ee1cf3c7 2020-04-09T14:53
  +0.001 XLM         IN         8ec0b206916ddd414db7eb0d1d74ef962cc14fb86a85cee924f01935614349df 2020-04-09T08:24
  +0.0123 XLM        IN         4ee461c717923f744d8aff9a9426946f3fe8941515aa5c9e863e023af298c285 2020-03-05T13:20
  +0.01 XLM          IN         9fa1cea94b11db5fc52d4077c0c69b56de5017df372fcb3356566eb3ca2361a8 2020-03-04T17:54
  +0 XLM             NONE       3294d225fb976707ba1680cbbb4b68ce512d99bc4c8abc2e89268716175c03a3 2020-03-04T09:27
  +0 XLM             NONE       52a6df1e2280cbde17943e52ec0dd799a96ae609b53bf81c4734703eeb27f43a 2020-03-04T09:22
  +2 XLM             IN         351362d4509d9bc3f33b39871755185b49ad5d14e45da16234438a35377e858a 2020-03-04T08:55
Stellar 3: 0 XLM (0ops) (GDYPMQMYW2JTLPWAUAHIDY3E4VHP5SGTFC5SMA45L7ZPOTHWQ2PHEW3E on 44'/148'/2') sep5#2 
OPERATIONS (0)
```

ksiva-ledger commented on src/families/stellar/api/horizon.js:
```diff
@@ -0,0 +1,261 @@
+
+const getSDKInstance = () => {
+  const baseAPIUrl = getEnv("API_STELLAR_HORIZON");
+  return new StellarSdk.Server(baseAPIUrl);
```
Yes, I did that because I saw an indication that we should use `getEnv` at runtime.
Otherwise I could have instantiated the SDK at the top of the file like before.

haammar-ledger replied:
I left a comment yesterday but apparently it got deleted...

You're right, but if you instantiate the SDK outside of a function I think it will just be a global variable but it will still be runtime.
You can check if the env variable can't be read when doing this, but if it works I think it's better to avoid creating a new SDK at every API call.

ksiva-ledger replied:
Done

ksiva-ledger commented:
> I tested with your changes and I don't have the error anymore on the 1st account 👏 but it is still here on the 2nd one.
> I think that account has something special... I don't remember what exactly I'll try to check tomorrow.

I'll test also with the Obelix seed in order to reproduce that.
I didn't had that issue with the simpler seed I used.

haammar-ledger commented on src/families/stellar/api/horizon.js:
```diff
@@ -0,0 +1,264 @@
+  return baseFee;
+};
+
+const getMinimumBalance = (account: RawAccount): BigNumber => {
+  const baseReserve = 0.5;
+  const numberOfEntries = account.subentry_count;
+
+  const minimumBalance = (2 + numberOfEntries) * baseReserve;
+
+  return parseCurrencyUnit(currency.units[0], minimumBalance.toString());
+};
```
This function doesn't use the API so you can move it to the `logic.js` file.

You can also do it for a few other functions in this file.

ksiva-ledger replied:
Done

haammar-ledger commented on cli/src/live-common-setup-base.js:
```diff
@@ -37,7 +37,7 @@ setSupportedCurrencies([
   "cosmos",
   "algorand",
-  "polkadot"
+  "polkadot",
```
Should remove this

haammar-ledger reviewed (approved):
:rocket:
